src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java

```diff
@@ -538,347 +511,128 @@ private ObjectMetadata syncPutObject(
         FileHashStoreUtility.ensureNotNull(pid, "pid", "tagObject");
         FileHashStoreUtility.ensureNotNull(cid, "cid", "tagObject");
-        FileHashStoreUtility.checkForEmptyString(pid, "pid", "tagObject");
-        FileHashStoreUtility.checkForEmptyString(cid, "cid", "tagObject");
-
-        synchronized (referenceLockedCids) {
-            while (referenceLockedCids.contains(cid)) {
-                try {
-                    referenceLockedCids.wait(TIME_OUT_MILLISEC);
-
-                } catch (InterruptedException ie) {
-                    String errMsg =
-                        "FileHashStore.tagObject - referenceLockedCids lock was interrupted while"
-                            + " waiting to tag pid: " + pid + " and cid: " + cid
-                            + ". InterruptedException: " + ie.getMessage();
-                    logFileHashStore.error(errMsg);
-                    throw new InterruptedException(errMsg);
-                }
-            }
-            logFileHashStore.debug(
-                "FileHashStore.tagObject - Synchronizing referenceLockedCids for cid: " + cid
-            );
-            referenceLockedCids.add(cid);
-        }
+        FileHashStoreUtility.checkForEmptyAndValidString(pid, "pid", "tagObject");
+        FileHashStoreUtility.checkForEmptyAndValidString(cid, "cid", "tagObject");
 
         try {
-            Path absPidRefsPath = getRealPath(pid, "refs", "pid");
-            Path absCidRefsPath = getRealPath(cid, "refs", "cid");
-
-            // Check that pid refs file doesn't exist yet
-            if (Files.exists(absPidRefsPath)) {
-                String errMsg = "FileHashStore.tagObject - pid refs file already exists for pid: "
-                    + pid + ". A pid can only reference one cid.";
-                logFileHashStore.error(errMsg);
-                throw new PidRefsFileExistsException(errMsg);
-
-            } else if (Files.exists(absCidRefsPath)) {
-                // Only update cid refs file if pid is not in the file
-                boolean pidFoundInCidRefFiles = isPidInCidRefsFile(pid, absCidRefsPath);
-                if (!pidFoundInCidRefFiles) {
-                    updateCidRefsFiles(pid, absCidRefsPath);
-                }
-                // Get the pid refs file
-                File pidRefsTmpFile = writePidRefsFile(cid);
-                File absPathPidRefsFile = absPidRefsPath.toFile();
-                move(pidRefsTmpFile, absPathPidRefsFile, "refs");
-                // Verify tagging process, this throws exceptions if there's an issue
-                verifyHashStoreRefsFiles(pid, cid, absPidRefsPath, absCidRefsPath);
-
-                logFileHashStore.info(
-                    "FileHashStore.tagObject - Object with cid: " + cid
-                        + " has been updated and tagged successfully with pid: " + pid
-                );
+            synchronizeObjectLockedCids(cid);
+            synchronizeReferenceLockedPids(pid);
```

Comment:
does it make more sense to move these locks into the `storeHashStoreRefsFiles` method, instead of managing them here? Ideally, they'd be as close to the synchronized code as possible, and having them in the other method would ensure nobody calls it in future without realizing they need to add a lock first

Reply:
I see where you're coming from - but when an unexpected exception occurs, I also need to revert the process via `unTagObject` and release the `cid` and `pid`. If we moved the synchronization process/locks to `storehashStoreRefsFiles(...)`, I feel that this process becomes a bit disjointed.

Since `storeHashStoreRefsFiles` is not part of the Public API, do you think it's safe for us to assume that adapter classes should implement `tagObject` in a thread safe manner, and not directly rely/use a protected method in `FileHashStore`? I'm not really sure, but if you feel strongly that we should move it - I am happy to.